### PR TITLE
Add relational storage layer for asset graph

### PR DESCRIPTION
**Pull request by mohavro**
## Summary
- introduce SQLAlchemy-backed ORM models, repository helpers, and an initial migration for assets, relationships, and regulatory events
- refactor AssetRelationshipGraph into a database-backed service that serves metrics and visualization data from the repository layer
- add integration tests that exercise CRUD flows through the repository and update existing tests and data loaders for the new persistence model

## Testing
- pytest tests/integration/test_repository.py *(fails: ModuleNotFoundError: No module named 'sqlalchemy' because the dependency cannot be installed in the execution environment)*

------
[Codex Task](https://chatgpt.com/codex/tasks/task_e_690e41adcb8c832498ce35d3c52b4424)

## Summary by Sourcery

Add a relational storage layer to the asset graph service by introducing SQLAlchemy ORM models, repository abstractions, and migrations; refactor the service to persist assets, relationships, and regulatory events in a database; and augment the test suite to validate end-to-end CRUD flows against an isolated SQLite store.

New Features:
- Introduce SQLAlchemy-backed ORM models and initial migration scripts for assets, relationships, and regulatory events
- Implement AssetGraphRepository for database CRUD operations and repository pattern
- Refactor AssetRelationshipGraph to use the relational database via session factory instead of in-memory structures
- Add integration tests for repository CRUD flows and update unit tests to use isolated SQLite databases

Enhancements:
- Optimize metrics computation and 3D visualization helpers to query persistent storage and leverage caching
- Remove legacy Supabase/PostgreSQL connector code and consolidate database setup into a unified SQLAlchemy module

Tests:
- Update fixtures to spin up isolated SQLite databases per test
- Enhance integration and unit tests to exercise the new persistence layer

### CODE REVIEW

**Comment by mohavro** [comment]
@codex review the following and make changes

Please address the comments from this code review: ## Overall Comments - The AssetRelationshipGraph class is doing persistence, analytics, and visualization all in one — consider splitting it into dedicated services or modules to improve separation of concerns and maintainability. - There’s duplicated logic for seeding and maintaining `_positions` in both visualization methods; extract this into a shared helper to reduce duplication. - Using lambdas with `_run_repository` everywhere introduces boilerplate — you could inject a repository instance or use a decorator to streamline CRUD operations and make the code more readable. ## Individual Comments ### Comment 1 <location> `src/logic/asset_graph.py:362-363` </location> <code_context> + self._positions = None + self._positions_known_asset_ids = [] + + def _positions_asset_ids_changed(self, asset_ids: List[str]) -> bool: + return asset_ids != self._positions_known_asset_ids + </code_context> <issue_to_address> **suggestion (bug_risk):** Asset ID comparison may be order-sensitive. Comparing unsorted asset_id lists may trigger unnecessary resets. Sorting both lists before comparison can prevent this issue. ```suggestion def _positions_asset_ids_changed(self, asset_ids: List[str]) -> bool: return sorted(asset_ids) != sorted(self._positions_known_asset_ids) ``` </issue_to_address> ### Comment 2 <location> `tests/integration/test_repository.py:42` </location> <code_context> + engine.dispose() + + +def test_asset_crud_flow(session): + repo = AssetGraphRepository(session) + </code_context> <issue_to_address> **suggestion (testing):** Consider adding tests for edge cases such as inserting assets with missing optional fields. Please include test cases for assets missing optional fields to confirm correct repository behavior. Suggested implementation: ```python def test_asset_crud_flow(session): repo = AssetGraphRepository(session) def test_insert_asset_missing_optional_fields(session): """ Test inserting an asset with missing optional fields to ensure repository handles it correctly. """ repo = AssetGraphRepository(session) # Example asset with only required fields; adjust field names as per your model asset_data = { "name": "Test Asset", # "description" is optional and omitted # "metadata" is optional and omitted } asset = repo.create(asset_data) assert asset.name == "Test Asset" assert getattr(asset, "description", None) is None assert getattr(asset, "metadata", None) is None ``` - Adjust the `asset_data` dictionary keys to match your actual asset model's required and optional fields. - If your repository's `create` method signature differs, update the call accordingly. - If your asset model uses different optional fields, update the assertions to match. </issue_to_address> ### Comment 3 <location> `tests/integration/test_repository.py:90` </location> <code_context> + assert "EQ1" not in assets + + +def test_relationship_and_event_crud_flow(session): + repo = AssetGraphRepository(session) + </code_context> <issue_to_address> **suggestion (testing):** Add tests for bidirectional relationships and duplicate relationship prevention. Add assertions to verify that bidirectional relationships are handled correctly and that duplicate relationships are prevented. Suggested implementation: ```python def test_relationship_and_event_crud_flow(session): repo = AssetGraphRepository(session) def test_bidirectional_relationships(session): repo = AssetGraphRepository(session) # Create two assets repo.create_asset("A", market_cap=1000) repo.create_asset("B", market_cap=2000) session.commit() # Add relationship A -> B repo.add_relationship("A", "B", "parent") session.commit() # Assert A has B as child relationships = repo.get_relationships("A") assert any(rel.target == "B" and rel.type == "parent" for rel in relationships) # Assert B has A as parent (bidirectional check, if supported) reverse_relationships = repo.get_relationships("B") assert any(rel.source == "A" and rel.type == "parent" for rel in reverse_relationships) def test_duplicate_relationship_prevention(session): repo = AssetGraphRepository(session) repo.create_asset("X", market_cap=500) repo.create_asset("Y", market_cap=600) session.commit() # Add relationship X -> Y twice repo.add_relationship("X", "Y", "linked") repo.add_relationship("X", "Y", "linked") session.commit() relationships = repo.get_relationships("X") # Assert only one relationship exists linked_rels = [rel for rel in relationships if rel.target == "Y" and rel.type == "linked"] assert len(linked_rels) == 1 ``` - If `get_relationships` does not support querying by source/target, you may need to adjust the assertions to match your repository API. - If the repository does not prevent duplicates by default, you may need to implement duplicate prevention logic in `add_relationship`. - If bidirectional relationships are not automatically created, you may need to explicitly add both directions and adjust the test accordingly. </issue_to_address> ### Comment 4 <location> `tests/unit/test_asset_graph.py:118` </location> <code_context> def test_same_sector_relationship(self): """Test that assets in the same sector are linked.""" - graph = AssetRelationshipGraph() </code_context> <issue_to_address> **suggestion (testing):** Add tests for edge cases such as assets with missing sector or mismatched asset classes. Please include tests for assets missing the sector attribute and for mismatched asset classes to verify robust handling of these scenarios. Suggested implementation: ```python def test_missing_sector_attribute(self): """Test that assets missing the sector attribute are handled gracefully.""" graph = AssetRelationshipGraph(database_url="sqlite:///:memory:") equity_with_sector = Equity( id="TECH1", name="Tech Equity 1", sector="Technology", asset_class="Equity" ) equity_missing_sector = Equity( id="TECH2", name="Tech Equity 2", sector=None, asset_class="Equity" ) graph.add_asset(equity_with_sector) graph.add_asset(equity_missing_sector) graph.build_relationships() # Only assets with sector should be linked by sector sector_links = graph.get_relationships_by_type("sector") assert all( equity_with_sector.id in link for link in sector_links ) assert all( equity_missing_sector.id not in link for link in sector_links ) def test_mismatched_asset_classes(self): """Test that assets with mismatched asset classes are not incorrectly linked.""" graph = AssetRelationshipGraph(database_url="sqlite:///:memory:") equity = Equity( id="EQ1", name="Equity 1", sector="Finance", asset_class="Equity" ) bond = Bond( id="BOND1", name="Bond 1", sector="Finance", asset_class="Bond" ) graph.add_asset(equity) graph.add_asset(bond) graph.build_relationships() # There should be no equity-to-bond sector links if logic is correct sector_links = graph.get_relationships_by_type("sector") for link in sector_links: assert not ( (equity.id in link) and (bond.id in link) ) def test_corporate_bond_relationship(self): """Test that corporate bonds are linked to their issuing equity.""" graph = AssetRelationshipGraph(database_url="sqlite:///:memory:") ``` - Ensure that the `Equity` and `Bond` classes, as well as the `add_asset`, `build_relationships`, and `get_relationships_by_type` methods, are available and correctly implemented in your codebase. - Adjust the asset creation and relationship logic if your actual implementation differs from the example shown. </issue_to_address> ### Comment 5 <location> `tests/unit/test_asset_graph.py:148` </location> <code_context> def test_corporate_bond_relationship(self): """Test that corporate bonds are linked to their issuing equity.""" - graph = AssetRelationshipGraph() </code_context> <issue_to_address> **suggestion (testing):** Add tests for bonds with missing issuer_id and for non-equity issuers. Include tests for bonds without issuer_id and for issuer_ids that do not match any equity asset to ensure relationships are not incorrectly created. </issue_to_address> ### Comment 6 <location> `tests/unit/test_api_main.py:286` </location> <code_context> def test_get_metrics_one_asset_no_relationships(self, client): """Metrics endpoint should handle graph with one asset and no relationships.""" - graph = AssetRelationshipGraph() </code_context> <issue_to_address> **suggestion (testing):** Consider adding tests for assets with relationships but missing regulatory events. Please add a test for the metrics endpoint with assets that have relationships but lack regulatory events, and check that the response fields are as expected. </issue_to_address> ### Comment 7 <location> `src/logic/asset_graph.py:392-398` </location> <code_context> if isinstance(asset1, Equity) and isinstance(asset2, Bond): if asset1.dividend_yield is not None and asset2.yield_to_maturity is not None: eps = 1e-6 num = abs(asset1.dividend_yield - asset2.yield_to_maturity) den = abs(asset1.dividend_yield) + abs(asset2.yield_to_maturity) + eps strength = max(0.0, 1.0 - num / den) relationships.append(("income_comparison", strength, True)) </code_context> <issue_to_address> **suggestion (code-quality):** Merge nested if conditions ([`merge-nested-ifs`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/Python/Default-Rules/merge-nested-ifs)) ```suggestion if isinstance(asset1, Equity) and isinstance(asset2, Bond) and (asset1.dividend_yield is not None and asset2.yield_to_maturity is not None): eps = 1e-6 num = abs(asset1.dividend_yield - asset2.yield_to_maturity) den = abs(asset1.dividend_yield) + abs(asset2.yield_to_maturity) + eps strength = max(0.0, 1.0 - num / den) relationships.append(("income_comparison", strength, True)) ``` <br/><details><summary>Explanation</summary>Too much nesting can make code difficult to understand, and this is especially true in Python, where there are no brackets to help out with the delineation of different nesting levels. Reading deeply nested code is confusing, since you have to keep track of which conditions relate to which levels. We therefore strive to reduce nesting where possible, and the situation where two `if` conditions can be combined using `and` is an easy win. </details> </issue_to_address>

File: src/logic/asset_graph.py [hunk 1]
```diff
@@ -1,72 +1,105 @@
+"""Asset relationship graph service backed by the relational database."""
+
+from __future__ import annotations
+
 from typing import Any, Dict, List, Optional, Tuple
```
[the remark this answers is not in the document]
Comment (mohavro) [reply src/logic/asset_graph.py]: @codex fix

In src/logic/asset_graph.py around line 5, the file imports Dict, List, and
Tuple from typing which is deprecated on Python 3.9+; replace those imports with
no typing imports for these collections and update all type hints in this file
to use native lowercase generics (e.g., dict[str, Asset], list[Asset],
tuple[str, int]) throughout the module. Ensure you keep Any and Optional imports
as needed (or replace Optional[T] with T | None if you prefer), adjust all
function signatures, variable annotations, and return types accordingly, and run
a quick type-check/static lint to catch any remaining old-typing usages.

**Review by mohavro — COMMENTED** [review]
pylint errors. Also several comments need reviews and fixes.

@codex review and fix
